addon/core/rdfa-processor.ts

```diff
@@ -6,41 +7,99 @@ import {
 } from '@lblod/ember-rdfa-editor/utils/_private/dom-helpers';
 import Datastore, {
   EditorStore,
+  SubAndContentPred,
 } from '@lblod/ember-rdfa-editor/utils/_private/datastore/datastore';
 import { Quad } from '@rdfjs/types';
 
-export type ExternalPropertyObject =
-  | {
-      type: 'literal';
-      rdfaId: string;
-    }
-  | {
-      type: 'resource';
-      resource: string;
-    };
-export type ExternalProperty = {
-  type: 'external';
-  predicate: string;
-  object: ExternalPropertyObject;
+export type SayTermType =
+  | 'NamedNode'
+  | 'Literal'
+  | 'LiteralNode'
+  | 'ResourceNode'
+  | 'ContentLiteral'
+  | 'BlankNode';
+
+export type SayRDFLiteral = Pick<
+  RDF.Literal,
+  'value' | 'termType' | 'language'
+> & {
+  datatype: Omit<RDF.NamedNode, 'equals'>;
 };
+export interface LiteralNodeObject {
+  termType: 'LiteralNode';
+  rdfaId: string;
+  datatype: Omit<RDF.NamedNode, 'equals'>;
+  language: string;
+}
+export interface ResourceNodeObject {
+  termType: 'ResourceNode';
+  value: string;
+}
+export interface ContentLiteralObject {
+  termType: 'ContentLiteral';
+  datatype: Omit<RDF.NamedNode, 'equals'>;
+  language: string;
+}
+export type PlainObject =
+  | SayRDFLiteral
+  | Omit<RDF.NamedNode, 'equals'>
+  | Omit<RDF.BlankNode, 'equals'>;
+export type NodeLinkObject = ResourceNodeObject | LiteralNodeObject;
 
-export type AttributeProperty = {
-  type: 'attribute';
+export type OutgoingTripleObject =
+  | PlainObject
+  | NodeLinkObject
+  | ContentLiteralObject;
+export interface LiteralTriple {
   predicate: string;
-  object: string;
-  datatype?: string;
-};
-export type ContentProperty = {
-  type: 'content';
+  object: SayRDFLiteral;
+}
+export interface NamedNodeTriple {
+  predicate: string;
+  object: Omit<RDF.NamedNode, 'equals'>;
+}
+export interface BlankNodeTriple {
   predicate: string;
+  object: Omit<RDF.BlankNode, 'equals'>;
+}
+export interface ResourceNodeTriple {
+  predicate: string;
+  object: ResourceNodeObject;
+}
+export interface LiteralNodeTriple {
+  predicate: string;
+  object: LiteralNodeObject;
+}
+export type ContentTriple = {
+  predicate: string;
+  object: ContentLiteralObject;
 };
-export type Property = AttributeProperty | ExternalProperty | ContentProperty;
 
-export type Backlink = {
+export type PlainTriple = LiteralTriple | NamedNodeTriple | BlankNodeTriple;
+export type LinkTriple = ResourceNodeTriple | LiteralNodeTriple;
+export type OutgoingTriple =
+  | LiteralTriple
+  | NamedNodeTriple
+  | BlankNodeTriple
+  | ResourceNodeTriple
+  | LiteralNodeTriple
+  | ContentTriple;
+
+export type IncomingResourceNodeTriple = {
+  termType: 'ResourceNode';
   subject: string;
   predicate: string;
```

Comment:
Wondering if we should also use `RDF.NamedNode` objects for the predicates/subjects.

Reply:
we could... it's just they can't be anything else so it's maybe overkill

Reply:
Fair enough, it doesn't really seem necessary in this case indeed.